### PR TITLE
Spike and slab2

### PR DESCRIPTION
**Pull request by mschauer**

### CODE REVIEW

File: scripts/zigzagboom1d.jl [hunk 1]
```diff
@@ -8,6 +8,7 @@ Random.seed!(1)
 # gradient of ϕ(x)
 ∇ϕ(x) = -π*sin(π*x) + x # (REPLACE IT WITH AUTOMATIC DIFFERENTIATION)
 
+∇ϕ(x) = x
```
Comment (mschauer) [inline scripts/zigzagboom1d.jl:11]: Undo this file

File: src/discretize.jl [hunk 2]
```diff
@@ -36,17 +45,18 @@ end
 Transform the output of the algorithm (a skeleton of points) to
 a trajectory. multi-dimensional version.
 """
-function discretize(x::Vector, Flow::Union{BouncyParticle, Boomerang}, dt0)
+function discretize(y::Vector, Flow::Union{BouncyParticle, Boomerang}, dt0)
     k = 1
+    x = deepcopy(y)
```
Comment (mschauer) [inline src/discretize.jl:50]: why is this necessary?

File: src/fact_samplers.jl [hunk 3]
```diff
@@ -45,7 +45,7 @@ from the upper upper bounding rates λᵢ(t) = max(a + b*t)^2. The factors `a` a
 can be function of the current position `x`, velocity `θ`, tuning parameter `c` and
 the Graph `G`
 """
-function ab(G, i, x, θ, c, Z::ZigZag)
+function ab(G, i, x, θ, c, Z::ZigZag, args...)
```
Comment (mschauer) [inline src/fact_samplers.jl:48]: ok

File: src/ss_non_fact.jl [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            tfrez[i] = t + freezing_time(x[i], θ[i])
+        end
+    end
+    tfrez
+end
+
+function refresh_sticky_vel!(θ,  θf, F::BouncyParticle)
+    for i in eachindex(θ)
+        if θ[i] == 0.0
+            θf[i] = abs(randn())sign(θf[i])
```
Comment (mschauer) [inline src/ss_non_fact.jl:15]:
```suggestion
            θf[i] = abs(randn())*sign(θf[i])
```

File: src/ZigZagBoomerang.jl [hunk 5]
```diff
@@ -7,7 +7,7 @@ include("common.jl")
 include("dynamics.jl")
 export ZigZag1d, Boomerang1d, ZigZag, FactBoomerang
 const LocalZigZag = ZigZag
-export LocalZigZag, BouncyParticle, Boomerang
+export LocalZigZag, BouncyParticle, Boomerang, StickyZigZag
```
Comment (mschauer) [inline src/ZigZagBoomerang.jl:10]: Do you define `StickyZigZag` somewhere?

File: src/ZigZagBoomerang.jl [hunk 6]
```diff
@@ -18,16 +18,26 @@ include("not_fact_samplers.jl")
...
 
 include("zigzagboom1d.jl")
 export pdmp, spdmp, eventtime, eventposition
 
+include("ss_fact.jl")
+export sspdmp
+
+
+include("ss_non_fact.jl")
+export sticky_pdmp
```
Comment (mschauer) [inline src/ZigZagBoomerang.jl:32]: inconsistent naming

File: src/deprecated_src/ss_fact_version1.jl [hunk 7]
```diff
@@ -0,0 +1,130 @@
+
```
Comment (mschauer) [inline src/deprecated_src/ss_fact_version1.jl:1]: do we need to `include` this file?

File: src/discretize.jl [hunk 8]
```diff
@@ -16,17 +16,26 @@ function discretize(x::Vector, Flow::Union{ZigZag1d, Boomerang1d}, dt0)
     dt = dt0
     while k < length(x)-1
         while clock + dt <= τ
-            clock, ξ, θ = move_forward(dt, clock, ξ, θ, Flow)
+            if θ == 0.0
```
Comment (mschauer) [inline src/discretize.jl:19]: What is this for?

File: src/deprecated_src/ss_fact_version1.jl (hunk 7 above)
Comment (SebaGraz) [reply src/deprecated_src/ss_fact_version1.jl:1]: Not at the moment. `ss_fact` is basically a duplicate of this file with minor changes. But I want to keep it so I can do a line by line comparison and see what are the changes.

File: src/ZigZagBoomerang.jl (hunk 6 above)
Comment (SebaGraz) [reply src/ZigZagBoomerang.jl:32]: Ok

File: src/ZigZagBoomerang.jl (hunk 5 above)
Comment (SebaGraz) [reply src/ZigZagBoomerang.jl:10]: No. It belonged to an old attempt of implementing the sticky zz

File: src/ss_non_fact.jl (hunk 4 above)
Comment (SebaGraz) [reply src/ss_non_fact.jl:15]: :+1:

File: src/discretize.jl (hunk 2 above)
Comment (SebaGraz) [reply src/discretize.jl:50]: Not really. I removed it

File: scripts/zigzagboom1d.jl (hunk 1 above)
Comment (SebaGraz) [reply scripts/zigzagboom1d.jl:11]: Done!

File: scripts/sticky/sticky_logitstic_sparse.jl [hunk 9]
```diff
@@ -0,0 +1,248 @@
+using ZigZagBoomerang
```
Comment (mschauer) [inline scripts/sticky/sticky_logitstic_sparse.jl:1]: File name typo

File: src/discretize.jl [hunk 10]
```diff
@@ -41,12 +50,12 @@ function discretize(x::Vector, Flow::Union{BouncyParticle, Boomerang}, dt0)
     _, ξ, θ = x[k]
     τ = x[k+1][1]
     clock = 0.0
-    Ω = trajectory([clock => ξ])
+    Ω = trajectory([clock => deepcopy(ξ)])
```
Comment (mschauer) [inline src/discretize.jl:53]: Here

File: scripts/sticky/sticky_logitstic_sparse.jl (hunk 9 above)
Comment (SebaGraz) [reply scripts/sticky/sticky_logitstic_sparse.jl]: Fixed

File: src/discretize.jl (hunk 8 above)
Comment (SebaGraz) [reply src/discretize.jl:19]: Just as a reference: the function `discretize` is implemented for the `ZigZag1d` and `Boomerang1d`. For the `Boomerang1d` with elliptic orbits not centred in 0, `move_forward` does not work when the particle is stuck in 0, since it would not stay in x,v = 0, 0.

File: src/discretize.jl (hunk 10 above)
Comment (SebaGraz) [reply src/discretize.jl:53]: This is an old version that would not work for the sticky Boomerang sampler. The current way to discretize the path is impelemnted in src/trace.jl (which probably will fail again for discretizing the path of the sticky Boomerang sampler). I think a `deepcopy` or at least a `copy` is necessary here.
